### ISSUE TITLE
Problems with decrypted secure variables getting extra characters.

### ISSUE THREAD

**Issue by danielksmith** [gocd/gocd#3930]
##### Issue Type
- Bug Report
##### Summary

I have the following two characters in a secure variable '£$'  when I run my pipeline these characters gain an extra character, 'Â£$', thereby invalidating a password.

###### Basic environment details

* Go Version: `Available in the footer`
* JAVA Version: `Available from http://<go-server-host>/go/about`
* OS: `Available from http://<go-server-host>/go/about`
* Browser vendor and version (if relevant): `Hopefully not Internet Explorer 7 :)`

##### Steps to Reproduce

1.  Open Notepad++ and make sure the encoding is utf-8
2.  Type £$
3.  Change the encoding to be ANSI
4. The text becomes Â£$

This is exactly what is happening in my GO pipeline. When I make my variable an Environment Variable and run my pipeline then its value is unchanged.

##### Expected Results
I would expect the values in secure variables to remain un-changed.

##### Actual Results
An extra character is added.

**Comment by ankitsri11** [comment]
Hi @danielksmith - I was trying to reproduce this on my local instance (GoCD v17.10 running on MacOS) but was unable to replicate. 

I tried below simple scenario - 

Secure Variable:
foo = bar£$

task -
echo $foo >> file 

Checked file on the agent box and can see the correct value "bar£$".

However, when I use in non secure variable, I get below:
[go] setting environment variable 'foo' to value 'bar�$'

Can you let me know the GoCD version and Operating system you are running on. Also, if possible can you share task configuration where you see this.

**Comment by danielksmith** [comment]
The GO Version is 'Go Version: 17.10.0' and the Build Agent's os is Windows Server 2016 Datacenter Edition

You say when you worked with a Secure Variable you didn't see this problem but you did when you used an Environment Variable. This is the opposite of the issue I encoutered. 

Dan.

**Comment by karlatkinson** [comment]
I've been playing around with this and the console log also fails to show (you can view it via the artifacts cruise-output). same as #3442 

Replicated on a new pipeline by adding one secure and one insecure variable.
Then just used a single task that executed a powershell script to write them to a file.

This is with gocd server running on ubuntu 14 with windows server 2016 as the agent os.

![image](https://user-images.githubusercontent.com/2210140/31897007-0d659490-b80d-11e7-9944-11018d379b97.png)

test_output.ps1
```
$insecure = Get-Item Env:insecure-thingy
$secure = Get-Item Env:secure-thingy

Add-Content -Encoding UTF8 output.txt $insecure.Value
Add-Content -Encoding UTF8 output.txt $secure.Value
```

**Comment by JasonCliftonZuto** [comment]
Do we have an update on this at all? We have issues storing connection strings as secure variables due to the use of the '£' sign

**Comment by ketan** [comment]
We'll try and get a fix in for the upcoming release.

**Comment by maheshp** [comment]
@danielksmith the server stores and reads the config in `UTF-8`, this looks like an issue where the encoding used on the agent is other than `UTF-8`. Can you please share the locale of the machine running the agent and server?

**Comment by karlatkinson** [comment]
@maheshp 

Server (ubuntu 14.04)

```
ubuntu@ip-xxx:~$ locale
LANG=en_US.UTF-8
LANGUAGE=
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=
```

Agent:
```
OS Name:                   Microsoft Windows Server 2016 Datacenter
OS Version:                10.0.14393 N/A Build 14393
System Locale:             en-us;English (United States)
Input Locale:              en-us;English (United States)
```

**Comment by maheshp** [comment]
@danielksmith and @JasonCliftonZuto we have pushed a fix for this issue. Can you please verify if the issue is fixed using an experimental build. The experimental installers can be downloaded from https://www.gocd.org/download  **Version: 17.12.0-5605** has the fixes.

@karlatkinson the console logs needs to be looked at as mentioned https://github.com/gocd/gocd/issues/3926#issuecomment-347520578

**Comment by karlatkinson** [comment]
@maheshp great, is the fix in both server and agent or can I just run the experimental agent?

**Comment by ketan** [comment]
The fix is in the server and agent. We recommend that you test this on a
test server and agent instance and not upgrade your production server to an
experimental build.

**Comment by karlatkinson** [comment]
@ketan Yeah, wasn't planning on that, was wondering if could just spin up a new agent though.
Will try it out with a new server and agent hopefully this week and feedback.

**Comment by rajiesh** [comment]
Verified this issue on 17.12.0 (5605-89a5ade1e9497ffc255dfa744476466c09beca06). The setup is - GoCD Server on MacOS and Agent on Windows server 2012. @jyotisingh did some debugging on the fix and it was observed that the server sends across the secure variable as is - with chars like £ intact - however while executing the command with the secure variable containing £ char, the command fails on the agent machine since the environment variable is not same as one received from the server(the £ char changes to different value on windows machine). Needs further investigation

**Comment by karlatkinson** [comment]
@ketan @maheshp confirmed that server and agent versions 17.12.0-5614 fix this issue for us.
Console log also show correct now. Not sure if this was a side-effect of fixing the encoding problem.

**Comment by karlatkinson** [comment]
@ketan @maheshp @rajiesh is there a reason this keeps slipping from releases?

**Comment by ketan** [comment]
A fix for this issue was merged in 17.12, but we forgot to close this issue.
